File: v2/src/main/mongo/indexes.ts
import type { MongoClient, Document, IndexSpecification } from 'mongodb';
import type {
  CreateIndexInput,
  IndexInfo,
  IndexKey,
  IndexStatRow,
} from '../../shared/schema.js';

export async function listIndexes(
  client: MongoClient,
  db: string,
  collection: string,
): Promise<IndexInfo[]> {
  const indexes = (await client.db(db).collection(collection).indexes()) as Document[];
  return indexes.map((idx) => {
    const keys: IndexKey[] = Object.entries(idx['key'] as Record<string, unknown>).map(
      ([field, dir]) => ({
        field,
        direction: dir as IndexKey['direction'],
      }),
    );
    return {
      name: String(idx['name']),
      keys,
      unique: Boolean(idx['unique']),
      sparse: Boolean(idx['sparse']),
      background: Boolean(idx['background']),
      ttlSeconds:
        typeof idx['expireAfterSeconds'] === 'number'
          ? (idx['expireAfterSeconds'] as number)
          : null,
      partialFilter: idx['partialFilterExpression']
        ? JSON.stringify(idx['partialFilterExpression'])
        : undefined,
    };
  });
}

export async function createIndex(
  client: MongoClient,
  db: string,
  collection: string,
  input: CreateIndexInput,
): Promise<string> {
  const spec: Record<string, IndexKey['direction']> = {};
  for (const k of input.keys) spec[k.field] = k.direction;
  const opts: Record<string, unknown> = {};
  if (input.name) opts['name'] = input.name;
  if (input.unique) opts['unique'] = true;
  if (input.sparse) opts['sparse'] = true;
  if (input.background) opts['background'] = true;
  if (input.ttlSeconds != null) opts['expireAfterSeconds'] = input.ttlSeconds;
  if (input.partialFilter) {
    try {
      opts['partialFilterExpression'] = JSON.parse(input.partialFilter);
    } catch (e) {
      throw new Error(
        `partialFilter is not valid JSON: ${(e as Error).message}`,
      );
    }
  }
  return client
    .db(db)
    .collection(collection)
    .createIndex(spec as IndexSpecification, opts);
}

export async function dropIndex(
  client: MongoClient,
  db: string,
  collection: string,
  name: string,
): Promise<void> {
  await client.db(db).collection(collection).dropIndex(name);
}

export async function indexStats(
  client: MongoClient,
  db: string,
  collection: string,
): Promise<IndexStatRow[]> {
  const rows = (await client
    .db(db)
    .collection(collection)
    .aggregate([{ $indexStats: {} }])
    .toArray()) as Document[];
  return rows.map((r) => ({
    name: String(r['name']),
    ops: Number((r['accesses'] as Document | undefined)?.['ops'] ?? 0),
    since:
      (r['accesses'] as Document | undefined)?.['since']
        ? String((r['accesses'] as Document)['since'])
        : null,
  }));
}

File: v3/src/main/kotlin/io/mex/mongo/Indexes.kt
package io.mex.mongo

import com.mongodb.client.MongoClient
import com.mongodb.client.model.IndexOptions
import org.bson.Document

data class IndexKey(val field: String, val direction: String)

data class IndexInfo(
    val name: String,
    val keys: List<IndexKey>,
    val unique: Boolean,
    val sparse: Boolean,
    val background: Boolean,
    val ttlSeconds: Long?,
    val partialFilter: String?,
)

data class CreateIndexInput(
    val keys: List<IndexKey>,
    val name: String? = null,
    val unique: Boolean = false,
    val sparse: Boolean = false,
    val background: Boolean = true,
    val ttlSeconds: Long? = null,
    val partialFilter: String? = null,
)

data class IndexStat(val name: String, val ops: Long)

fun listIndexes(client: MongoClient, db: String, coll: String): List<IndexInfo> {
    val docs = client.getDatabase(db).getCollection(coll).listIndexes().toList()
    return docs.map { d ->
        @Suppress("UNCHECKED_CAST")
        val keyMap = (d["key"] as Document).toMap()
        IndexInfo(
            name = d.getString("name"),
            keys = keyMap.map { (k, v) -> IndexKey(k, v.toString()) },
            unique = d.getBoolean("unique", false),
            sparse = d.getBoolean("sparse", false),
            background = d.getBoolean("background", false),
            ttlSeconds = (d["expireAfterSeconds"] as? Number)?.toLong(),
            partialFilter = (d["partialFilterExpression"] as? Document)?.toJson(),
        )
    }
}

fun createIndex(client: MongoClient, db: String, coll: String, input: CreateIndexInput): String {
    val spec = Document()
    for (k in input.keys) {
        val v: Any = k.direction.toIntOrNull() ?: k.direction
        spec[k.field] = v
    }
    val opts = IndexOptions()
    input.name?.let { opts.name(it) }
    if (input.unique) opts.unique(true)
    if (input.sparse) opts.sparse(true)
    if (input.background) opts.background(true)
    input.ttlSeconds?.let { opts.expireAfter(it, java.util.concurrent.TimeUnit.SECONDS) }
    input.partialFilter?.takeIf { it.isNotBlank() }?.let {
        opts.partialFilterExpression(Document.parse(it))
    }
    return client.getDatabase(db).getCollection(coll).createIndex(spec, opts)
}

fun dropIndex(client: MongoClient, db: String, coll: String, name: String) {
    client.getDatabase(db).getCollection(coll).dropIndex(name)
}

fun indexStats(client: MongoClient, db: String, coll: String): List<IndexStat> {
    return runCatching {
        client.getDatabase(db).getCollection(coll)
            .aggregate(listOf(Document("\$indexStats", Document())))
            .toList()
            .map {
                IndexStat(
                    name = it.getString("name"),
                    ops = ((it["accesses"] as? Document)?.get("ops") as? Number)?.toLong() ?: 0L,
                )
            }
    }.getOrDefault(emptyList())
}
